### PR TITLE
implemented some more archivist features, also a bit of refactoring

### PR DESCRIPTION
**Pull request by AlmirKadric**

### CODE REVIEW

File: Mage/Archivist/Archivist.cs [hunk 1]
```diff
@@ -240,35 +424,27 @@ public class Archivist : EventEmitter<JObject> {
...
-						_cache.Add (cacheKeyName, newValue);
-						responseResult = (JToken)newValue.data;
-					}
+					// Set value to cache
+					SetCacheValue(topicValue);
 
 					// Add value to response
 					string responseKey = realQueryKeys[cacheKeyName];
-					responseObject.Add(responseKey, responseResult);
+					responseObject.Add(responseKey, GetCacheValue(cacheKeyName).data);
```
Comment (nullorvoid) [inline Mage/Archivist/Archivist.cs:447]: GetCacheValue can return null, will this explode? 

Edit: Ah its in a try catch, I see you are setting it then return it here but in a raw get.

**Comment by christopherobin** [comment]
Looks good to me

**Comment by nullorvoid** [comment]
LGTM

File: Mage/Archivist/Archivist.cs (hunk 1 above)
Comment (christopherobin) [inline Mage/Archivist/Archivist.cs:447]: SetCacheValue is called right above it so it should never return null I think, unless the expiration was set to 0?

File: Mage/Archivist/Archivist.cs [hunk 2]
```diff
@@ -54,8 +76,166 @@ public class Archivist : EventEmitter<JObject> {
...
+
+	// Return cache dictionary
+	public Dictionary<string, VaultValue> GetCache() {
+		return _cache;
+	}
+
+
+	// Clear out the cache entirely
+	public void ClearCache() {
+		_cache = new Dictionary<string, VaultValue>();
```
Comment (ronkorving) [inline Mage/Archivist/Archivist.cs:123]:
``` cs
_cache.Clear();
```

is a bit more elegant imho (no type redefinition).

File: Mage/Archivist/Archivist.cs [hunk 3]
```diff
@@ -54,8 +76,166 @@ public class Archivist : EventEmitter<JObject> {
...
+		}
+		
+		// Do delete
+		cacheValue.Del();
+		
+		// Emit touch event
+		this.emit(topic + ":del", cacheValue);
+	}
+
+	private void ValueTouch(string topic, JObject index, int? expirationTime) {
```
Comment (ronkorving) [inline Mage/Archivist/Archivist.cs:203]: Why is expirationTime optional?

File: Mage/Archivist/VaultValue.cs [hunk 4]
```diff
@@ -16,29 +16,70 @@ public class VaultValue {
...
 
-		_mediaType = valueObject["value"]["mediaType"].ToString();
+	// TODO: implement multiple media-types and encoding
+	public void SetData(string mediaType, JToken data) {
+		// Detect media type
+		_mediaType = mediaType;
 
-		// TODO: implement multiple media-types
-		_data = Tome.Conjure(JToken.Parse(valueObject["value"]["data"].ToString()));
+		// Set data based on media type
```
Comment (ronkorving) [inline Mage/Archivist/VaultValue.cs:39]: I think that old TODO is still valid, no?

Comment (AlmirKadric) [inline Mage/Archivist/VaultValue.cs:39]: it's still there, its been moved to line 34

File: Mage/Archivist/Archivist.cs (hunk 3 above)
Comment (AlmirKadric) [inline Mage/Archivist/Archivist.cs:203]: the backend will not provide an expiration time (undefined) if there is no TTL on the data

File: Mage/Archivist/Archivist.cs (hunk 2 above)
Comment (AlmirKadric) [inline Mage/Archivist/Archivist.cs:123]: i wonder if clear is thread safe
the issue here is that clear could be deleting record by record which can break in threading
however if we create a new object and change the reference, we're guaranteed a new object

Comment (ronkorving) [inline Mage/Archivist/Archivist.cs:123]: Fair enough. You may want to add a comment with that reasoning, as someone might "optimize" your code here.

File: Mage/Archivist/Archivist.cs (hunk 3 above)
Comment (ronkorving) [inline Mage/Archivist/Archivist.cs:203]: How could touch not have an expirationTime?

**Comment by AlmirKadric** [comment]
above comments considered, this is good to go
any required changes can become an issue from here on out

File: Mage/Archivist/VaultValue.cs (hunk 4 above)
Comment (ronkorving) [inline Mage/Archivist/VaultValue.cs:39]: Ah, right :+1:

File: Mage/Archivist/Archivist.cs (hunk 2 above)
Comment (AlmirKadric) [inline Mage/Archivist/Archivist.cs:123]: will add in next PR

File: Mage/Archivist/Archivist.cs (hunk 3 above)
Comment (AlmirKadric) [inline Mage/Archivist/Archivist.cs:203]: ah touch might have an expiration time, but also might remove it as well no?
i.e. there was a ttl, but the ttl was replaced with a non-expiring piece of data

Comment (ronkorving) [inline Mage/Archivist/Archivist.cs:203]: Hmmm, good question. I don't have a direct answer to that. Feel free to look into that edge-case.

**Pull request merged**